### PR TITLE
Windows: Add response files support for long command lines

### PR DESCRIPTION
**Pull request by kinke**
With the Ninja backend and D/C[++] languages.

### CODE REVIEW

**Comment by kinke** [comment]
I've seen that unittests in `dcompile.d` aren't tested; I'd like to add some tests wrt. response file parsing...

**Comment by atilaneves** [comment]
What unit tests in dcompile?

File: payload/reggae/backend/ninja.d [hunk 1]
```diff
@@ -70,6 +70,28 @@ private bool hasDepFile(in CommandType type) @safe pure nothrow {
     return type == CommandType.compile || type == CommandType.compileAndLink;
 }
 
+private string[] initializeRuleParamLines(in Language language, in string command) @safe pure {
+    version(Windows) {
+        // On Windows, the max command line length is ~32K.
+        // Make ninja use a response file for all D/C[++] rules.
+        if (language == Language.D || language == Language.C || language == Language.Cplusplus) {
```
Comment (atilaneves) [inline payload/reggae/backend/ninja.d:77]: `if(language.among(Language.D, ...))`

File: payload/reggae/dcompile.d [hunk 2]
```diff
@@ -110,6 +128,33 @@ private string[] mapToLdcOptions(in string[] compArgs) @safe pure {
...
+private auto invokeCompiler(in string[] args, in string objFile) @safe {
+    version(Windows) {
+        static string quoteArgIfNeeded(string a) {
+            return !a.canFind(' ') ? a : `"` ~ a.replace(`"`, `\"`) ~ `"`;
+        }
+
+        const rspFileContent = args[1..$].map!quoteArgIfNeeded.join("\n");
+
+        // max command-line length (incl. args[0]) is ~32,767 on Windows
+        /*if (rspFileContent.length > 32_000)*/ {
```
Comment (atilaneves) [inline payload/reggae/dcompile.d:140]: ?

File: tests/it/runtime/dub.d [hunk 3]
```diff
@@ -18,7 +18,12 @@ unittest {
         writelnUt("\n\nReggae output:\n\n", runReggae("-b", "ninja").lines.join("\n"), "-----\n");
         shouldExist("reggaefile.d");
         auto output = ninja.shouldExecuteOk;
-        output.shouldContain("-debug -g");
+
+        version(Windows) {
+            // args in response file
```
Comment (atilaneves) [inline tests/it/runtime/dub.d:23]: Shouldn't we test that the args are actually in the response file on Windows?

File: payload/reggae/dcompile.d (hunk 2 above)
Comment (kinke) [reply payload/reggae/dcompile.d:140]: Just temporarily commented out to have dcompile's output response file tested via CI with both DMD and LDC - there are apparently no long cmdlines in the testsuite.

**Comment by kinke** [comment]
> What unit tests in dcompile?

The ones I would have liked to add. ;) - Guess I'd have to extract the response file parsing to its own file and add according tests in the `tests/ut` directory...

File: payload/reggae/dcompile.d (hunk 2 above)
Comment (atilaneves) [reply payload/reggae/dcompile.d:140]: Ah, ok. Please revert.

File: tests/it/runtime/dub.d (hunk 3 above)
Comment (kinke) [reply tests/it/runtime/dub.d:23]: That would IMO be a PITA, even require to run ninja with `-d keeprsp` to keep the input response file. `tests/ut/ninja.d` tests that `rspfile_contents` contains `$flags` (and all the other compiler args in the normal cmdline), and the enabled output check on Posix makes sure the expected flags make it into the normal cmdline.

**Comment by kinke** [comment]
Okay, I've added a few unittests.

**Review by atilaneves — APPROVED** [review]

**Pull request merged**